AMDGPU/GlobalISel: AMDGPURegBankSelect

petar-avramovic opened the pull request:
Assign register banks to virtual registers. Does not use generic
RegBankSelect. After register bank selection all register operand of
G_ instructions have LLT and register banks exclusively. If they had
register class, reassign appropriate register bank.

Assign register banks using machine uniformity analysis:
Sgpr - uniform values and some lane masks
Vgpr - divergent, non S1, values
Vcc  - divergent S1 values(lane masks)

AMDGPURegBankSelect does not consider available instructions and, in
some cases, G_ instructions with some register bank assignment can't be
inst-selected. This is solved in RegBankLegalize.

Exceptions when uniformity analysis does not work:
S32/S64 lane masks:
- need to end up with sgpr register class after instruction selection
- In most cases Uniformity analysis declares them as uniform
  (forced by tablegen) resulting in sgpr S32/S64 reg bank
- When Uniformity analysis declares them as divergent (some phis),
  use intrinsic lane mask analyzer to still assign sgpr register bank
temporal divergence copy:
- COPY to vgpr with implicit use of $exec inside of the cycle
- this copy is declared as uniform by uniformity analysis
- make sure that assigned bank is vgpr
Note: uniformity analysis does not consider that registers with vgpr def
are divergent (you can have uniform value in vgpr).
- TODO: implicit use of $exec could be implemented as indicator
  that instruction is divergent

petar-avramovic commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112863?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#112864** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112864?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#112863** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112863?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#112862** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112862?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @petar-avramovic and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+  return true;
+}
+
+void setRB(MachineInstr &MI, MachineOperand &DefOP, MachineIRBuilder B,
```
```suggestion
void setRBDef(MachineInstr &MI, MachineOperand &DefOp, MachineIRBuilder B,
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+                                  : std::next(MI.getIterator()));
+    B.buildCopy(Reg, NewReg);
+
+    // The problem was discoverd for uniform S1 that was used as both
```
```suggestion
    // The problem was discovered for uniform S1 that was used as both
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+    // - lane-mask(vcc) use was by si_if, this use is divergent and requires
+    //   non-trivial sgpr-S1-to-vcc copy. But pre-inst-selection of si_if sets
+    //   sreg_64_xexec(S1) on def of uniform S1 making it lane-mask.
+    // - the regular regular sgpr S1(uniform) instruction is now broken since
```
```suggestion
    // - the regular sgpr S1(uniform) instruction is now broken since
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+    // - the regular regular sgpr S1(uniform) instruction is now broken since
+    //   it uses sreg_64_xexec(S1) which is divergent.
+
+    // "Clear" reg classes from uses on generic instructions and but register
```
```suggestion
    // "Clear" reg classes from uses on generic instructions and put register
```

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+    for (auto &UseMI : MRI.use_instructions(Reg)) {
+      if (shouldRBSelect(UseMI)) {
+        for (MachineOperand &Op : UseMI.operands()) {
+          if (Op.isReg() && Op.isUse() && Op.getReg() == Reg)
```
Do we really need to check isUse?

rovka commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+
+    auto &MBB = *MI.getParent();
+    B.setInsertPt(MBB, MI.isPHI() ? MBB.getFirstNonPHI()
+                                  : std::next(MI.getIterator()));
```
Can you use skipPHIsAndLabels(std:next...) here too?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -69,3 +72,38 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+void IntrinsicLaneMaskAnalyzer::initLaneMaskIntrinsics(MachineFunction &MF) {
+  for (auto &MBB : MF) {
+    for (auto &MI : MBB) {
+      if (MI.getOpcode() == AMDGPU::G_INTRINSIC &&
+          MI.getOperand(MI.getNumExplicitDefs()).getIntrinsicID() ==
+              Intrinsic::amdgcn_if_break) {
```
`cast<GIntrinsic>(MI).getIntrinsicID() == Intrinsic::amdgcn_if_break`

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -69,3 +72,38 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+
+void IntrinsicLaneMaskAnalyzer::findLCSSAPhi(Register Reg) {
+  S32S64LaneMask.insert(Reg);
+  for (auto &LCSSAPhi : MRI.use_instructions(Reg)) {
```
```suggestion
  for (const MachineInstr &LCSSAPhi : MRI.use_instructions(Reg)) {
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+  if (isTargetSpecificOpcode(MI.getOpcode()) && !MI.isPreISelOpcode())
+    return false;
+
+  if (MI.getOpcode() == AMDGPU::PHI || MI.getOpcode() == AMDGPU::IMPLICIT_DEF)
+    return false;
```
These should have failed isPreISelOpcode

petar-avramovic replied:
I copied that from existing regbankselect. MI.isPreISelOpcode() || MI.isCopy() also works

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+  if (MI.getOpcode() == AMDGPU::PHI || MI.getOpcode() == AMDGPU::IMPLICIT_DEF)
+    return false;
+
+  if (MI.isInlineAsm())
```
These should have failed isPreISelOpcode

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+      if (!Op.isReg())
+        continue;
+      Register Reg = Op.getReg();
+      if (Reg == AMDGPU::EXEC) {
```
What about wave32

petar-avramovic replied:
I used exec regardless of wave32/64 should I switch to TRI.getExec()?
```
    B.buildInstr(AMDGPU::COPY, {VgprReg}, {Reg})
        .addUse(AMDGPU::EXEC, RegState::Implicit);
```

arsenm replied:
Yes (although I suppose there's a question of how to interpret reading only exec_lo on wave64 and other edge cases)

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+
+Register getVReg(MachineOperand &Op) {
+  if (!Op.isReg())
+    return 0;
```
Use explicit Register()

tschuett replied:
`std::nullopt`

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+
+  Register Reg = Op.getReg();
+  if (!Reg.isVirtual())
+    return 0;
```
Use explicit Register()

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const RegisterBankInfo &RBI = *MF.getSubtarget().getRegBankInfo();
+
+  MachineIRBuilder B(MF);
```
Should initialize the CSE info and other analyses

petar-avramovic replied:
Do we need CSE bilder here? This pass only sets register banks, and in some cases builds copies but dont think we need CSE for them.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+      getAnalysis<MachineUniformityAnalysisPass>().getUniformityInfo();
+  AMDGPU::IntrinsicLaneMaskAnalyzer ILMA(MF);
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const RegisterBankInfo &RBI = *MF.getSubtarget().getRegBankInfo();
```
getSubtarget<GCNSubtarget> and use the target specific instance

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+          setRB(MI, DefOP, B, MRI, RBI.getRegBank(AMDGPU::SGPRRegBankID));
+        } else {
+          if (MRI.getType(DefReg) == LLT::scalar(1))
+            setRB(MI, DefOP, B, MRI, RBI.getRegBank(AMDGPU::VCCRegBankID));
```
Can you directly use the pointer to the const regbank struct?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+        continue;
+
+      for (MachineOperand &UseOP : MI.uses()) {
+        Register UseReg = getVReg(UseOP);
```
I think getVReg isn't helping readability, it's more canonical to just directly have the isReg + isVirtual check here

tschuett replied:
Please use std::optional.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+      if (MI.isCopy())
+        continue;
+
+      for (MachineOperand &UseOP : MI.uses()) {
```
What about the defs?

petar-avramovic replied:
Previous for loop assigned RegBanks to all defs.
This loop prepares uses for RBLegalize to have register banks only.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+  }
+}
+
+void setRBUse(MachineInstr &MI, MachineOperand &UseOP, MachineIRBuilder B,
```
static. Also don't pass MachineIRBuilder by value

tschuett replied:
member function

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+    }
+
+  } else {
+    MRI.setRegBank(Reg, RB);
```
Do you need to call the observer?

petar-avramovic replied:
Since this is our pass I felt there was no need to complicate it with observers

arsenm reviewed (commented):
Don't forget about AGPRs

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+//   instruction is divergent
+bool isTemporalDivergenceCopy(Register Reg, MachineRegisterInfo &MRI) {
+  MachineInstr *MI = MRI.getVRegDef(Reg);
+  if (MI->getOpcode() == AMDGPU::COPY) {
```
Early exit on non-copy and reduce indent

tschuett replied:
member function and use instance variables, i.e., MRI.

tschuett commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
+  return false;
+}
+
+Register getVReg(MachineOperand &Op) {
```
std::optional<Register>

tschuett replied:
And probably something like `tryGetVReg`. It is fallible.

tschuett commented on llvm/lib/Target/AMDGPU/AMDGPURBSelect.cpp:
```diff
@@ -63,4 +70,189 @@ char &llvm::AMDGPURBSelectID = AMDGPURBSelect::ID;
 FunctionPass *llvm::createAMDGPURBSelectPass() { return new AMDGPURBSelect(); }
 
-bool AMDGPURBSelect::runOnMachineFunction(MachineFunction &MF) { return true; }
+bool shouldRBSelect(MachineInstr &MI) {
```
Why free-standing functions, when it is **your**. register bank select pass?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -69,3 +72,37 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+  initLaneMaskIntrinsics(MF);
+}
+
+bool IntrinsicLaneMaskAnalyzer::isS32S64LaneMask(Register Reg) {
```
const (but really I don't see how this wrapper helps readability"

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -69,3 +72,37 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+void IntrinsicLaneMaskAnalyzer::initLaneMaskIntrinsics(MachineFunction &MF) {
+  for (auto &MBB : MF) {
+    for (auto &MI : MBB) {
+      if (isa<GIntrinsic>(MI) &&
+          cast<GIntrinsic>(MI).getIntrinsicID() == Intrinsic::amdgcn_if_break) {
```
dyn_cast instead isa+cast

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +73,215 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+    // of another instruction. Maybe cross bank copy was required so we insert a
+    // copy that can be removed later. This simplifies post regbanklegalize
+    // combiner and avoids need to special case some patterns.
+    if (MRI.getRegClassOrNull(Reg)) {
```
Early exit in the simple case and reduce indent

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +73,215 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+    if (MRI.getRegClassOrNull(Reg)) {
+      LLT Ty = MRI.getType(Reg);
+      Register NewReg = MRI.createVirtualRegister({RB, Ty});
+      DefOP.setReg(NewReg);
```
Does this need to call the observer

petar-avramovic replied:
Why? I intend for new regbankselect be simple and not use observers.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +73,215 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+        if (shouldRegBankSelect(UseMI)) {
+          for (MachineOperand &Op : UseMI.operands()) {
+            if (Op.isReg() && Op.getReg() == Reg)
+              Op.setReg(NewReg);
```
Modifying use list while in the use_instructions is hazardous?

petar-avramovic replied:
Yes, this was a bug.
make_early_inc_range also works but it might be assuming how MRI keeps track of use instructions internally

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +73,215 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+    }
+  }
+
+  std::optional<Register> tryGetVReg(MachineOperand &Op) {
+    if (!Op.isReg())
+      return std::nullopt;
```
Might as well just use Register()? Do we have a specialization for optional on Register?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +73,215 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
           MachineFunctionProperties::Property::FailedISel))
     return false;
+
+  MachineIRBuilder B(MF);
```
Should init with the CSEInfo etc.

petar-avramovic replied:
Added, but I don't see a reason for it. Here we only set register banks and insert a few copies, there is nothing for CSE to find. Legalizer part actually makes use of CSE often.

arsenm replied:
It's not about this pass, it's preserving it through the compilation flow

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,232 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+      return false;
+
+    for (auto Op : MI->implicit_operands()) {
+      if (!Op.isReg())
+        continue;
```
This is illegal for a copy

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,232 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+    if (!MI->isCopy())
+      return false;
+
+    for (auto Op : MI->implicit_operands()) {
```
Don't know why this is only looking at the implicit operands, and not the only real operand (which I would expect is the only one that matters). Just hardcode looking at the first operand?

petar-avramovic replied:
There is nothing useful in real operands, it is a simple sgpr to vgpr COPY which is quite common. What makes that COPY special is implicit exec, for example:
`%21:vgpr_32(s32) = COPY %8:sgpr(s32), implicit $exec_lo`
implicit exec is there to stop other passes from moving the COPY outside of the loop.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,232 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+    return false;
+  }
+
+  void setRBDef(MachineInstr &MI, MachineOperand &DefOP,
```
Avoid RB abbreviation?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,232 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+        }
+      }
+    }
+    for (MachineOperand *Op : RegUsesOnGInstrs) {
```
Avoid collecting all the instructions to change. Can you either use the early_inc_range (or better yet, avoid the use list walk)?)

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,232 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+  }
+
+  void assignBanksOnDefs(MachineInstr &MI) {
+    if (!shouldRegBankSelect(MI))
+      return;
```
Check this as a pre-condition to enter?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -69,3 +72,37 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+        findLCSSAPhi(MI.getOperand(0).getReg());
+      }
+
+      if (MI.getOpcode() == AMDGPU::SI_IF ||
+          MI.getOpcode() == AMDGPU::SI_ELSE) {
+        findLCSSAPhi(MI.getOperand(0).getReg());
+      }
```
I don't understand why you are mixing matching the intrinsic form of if.break above, but the selected pseudos for if and else

petar-avramovic replied:
Consequence of what legalizer does, si.if and si.else are inst-selected to SI_IF and SI_ELSE in AMDGPULegalizerInfo::legalizeIntrinsic, if.break is still intrinsic for reg bank selection

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,208 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+    if (!MI->isCopy() || MI->getNumImplicitOperands() != 1)
+      return false;
+
+    return MI->implicit_operands().begin()->getReg() == TRI.getExec();
```
Just hardcode the operand index

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +81,208 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+  // are divergent (you can have uniform value in vgpr).
+  // - TODO: implicit use of $exec could be implemented as indicator that
+  //   instruction is divergent
+  bool isTemporalDivergenceCopy(Register Reg) {
```
Special casing a copy with an implicit exec operand doesn't feel hermetic. We already have ugly hacks to insert these copies at one point, and it's not a wholistic strategy. There are ample opportunities to lose the operand, such as introducing new derived operations which do not have the use 

Where are these exec operands getting inserted? Should we have a different pseudo instead?

petar-avramovic replied:
> Where are these exec operands getting inserted? Should we have a different pseudo instead?

In temporal-divergence lowering, upcoming change in AMDGPUGlobalISelDivergenceLowering, I will just remove this check in this patch.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -0,0 +1,271 @@
+    }
+  }
+
+  // At this point all virtual registers have register class or bank
```
Can you do this in one loop with a post order traversal?

petar-avramovic replied:
Merged two loops into one. Order does not matter, could go
  for (unsigned i = 0; i < MRI.getNumVirtRegs(); ++i) {
    Register Reg = Register::index2VirtReg(i);
but still needs to search for def instructions and lists of uses, don't know if that might be better option

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankSelect.cpp:
```diff
@@ -66,9 +78,183 @@ FunctionPass *llvm::createAMDGPURegBankSelectPass() {
+      // or bank. If there is neither on vreg in def operand, assign bank.
+      if (MI.isCopy()) {
+        Register DefReg = MI.getOperand(0).getReg();
+        assert(!MRI.getRegBankOrNull(DefReg));
+        if (!MRI.getRegClassOrNull(DefReg))
```
Either the assert or the if are redundant

arsenm replied:
ping?

petar-avramovic replied:
Sorry, missed that, will fix in next update

petar-avramovic replied:
Small rework, still want to have both checks. Vregs on def before assigning banks should have either nothing or register class. Assert is there to check that there is no bank already assigned

petar-avramovic commented:
There is something wrong with one loop approach, multiple tests are failing. They all have temporal divergent values so I assume that is the issue. Need to investigate

petar-avramovic commented:
All fixed (error was in upcoming patches). Ready for review

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -68,3 +72,37 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+  }
+}
+
+void IntrinsicLaneMaskAnalyzer::findLCSSAPhi(Register Reg) {
+  S32S64LaneMask.insert(Reg);
+  for (const MachineInstr &LCSSAPhi : MRI.use_instructions(Reg)) {
+    if (LCSSAPhi.isPHI())
+      S32S64LaneMask.insert(LCSSAPhi.getOperand(0).getReg());
+  }
+}
```
I'm curious: How sure are you/we that there can never be "nested" phis here, perhaps due to breaking out of nested loops?

petar-avramovic replied:
IntrinsicLaneMaskAnalyzer is very very trivial. It only handles LCSSA phis and if.break phis.

Control flow intrinsic should only be used by other control flow intrinsics. But since we run LCSSA they can also be used by lcssa-phis.

findLCSSAPhi is really meant to cover intrinsics used for control flow that are affected by LCSSA pass.
`%64:_(s32) = G_PHI %15(s32), %bb.3` is lane mask. Uniformity analysis says it is divergent. If it was S1 we would be fine but since it is S32 we need to special case put it to sgpr.

```
    %15:sreg_32_xm0_xexec(s32) = G_INTRINSIC intrinsic(@llvm.amdgcn.if.break), %45(s1), %14(s32)
    SI_LOOP %15(s32), %bb.1, implicit-def $exec, implicit-def $scc, implicit $exec
    G_BR %bb.6

  bb.6:
    %64:_(s32) = G_PHI %15(s32), %bb.3
    G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.amdgcn.end.cf), %64(s32)
    S_ENDPGM 0
```

arsenm replied:
Is this just working around a bug in uniformity analysis? We already special case amdgcn_if and else in isAlwaysUniform. Did this just miss if.break?

petar-avramovic replied:
if.break is already forced uniform.
Here we special case lcssa-phi lane masks.
Phi is divergent because it uses value(uniform if.break in this case) outside cycle with the divergent exit, but that phi is lane mask so it needs to end up in sgpr.

nhaehnle reviewed (approved):
I do have a comment/question, but I think we're at the point where it just makes sense to make progress and land this.

petar-avramovic committed:
AMDGPU/GlobalISel: AMDGPURegBankSelect

Assign register banks to virtual registers. Does not use generic
RegBankSelect. After register bank selection all register operand of
G_ instructions have LLT and register banks exclusively. If they had
register class, reassign appropriate register bank.

Assign register banks using machine uniformity analysis:
Sgpr - uniform values and some lane masks
Vgpr - divergent, non S1, values
Vcc  - divergent S1 values(lane masks)

AMDGPURegBankSelect does not consider available instructions and, in
some cases, G_ instructions with some register bank assignment can't be
inst-selected. This is solved in RegBankLegalize.

Exceptions when uniformity analysis does not work:
S32/S64 lane masks:
- need to end up with sgpr register class after instruction selection
- In most cases Uniformity analysis declares them as uniform
  (forced by tablegen) resulting in sgpr S32/S64 reg bank
- When Uniformity analysis declares them as divergent (some phis),
  use intrinsic lane mask analyzer to still assign sgpr register bank
temporal divergence copy:
- COPY to vgpr with implicit use of $exec inside of the cycle
- this copy is declared as uniform by uniformity analysis
- make sure that assigned bank is vgpr
Note: uniformity analysis does not consider that registers with vgpr def
are divergent (you can have uniform value in vgpr).
- TODO: implicit use of $exec could be implemented as indicator
  that instruction is divergent
